automata/base/animation.py

```diff
@@ -0,0 +1,254 @@
+from typing import Optional, TypeVar
+
+import manim
+import pygraphviz as pgv
+
+_POINTS_IN_INCH = 72
+
+
+class Animate:
+    """The `Animate` class is the config class to set the behavior of animations."""
+
+    DEFAULT_COLOR = manim.WHITE
+    HIGHLIGHT_COLOR = manim.RED
+    M = TypeVar("M", bound=manim.Mobject)
+
+    @classmethod
+    def default_init(cls, mobject_class: type[M]) -> type[M]:
+        return partial(mobject_class, color=cls.DEFAULT_COLOR)  # type: ignore
+
+    @classmethod
+    def to_default_color(cls, mobject: manim.VMobject) -> manim.ApplyMethod:
+        """
+        Parameters
+        ----------
+        mobject : VMobject
+            The mobject to change to `DEFAULT_COLOR`, or cancel the highlight.
+
+        Returns
+        -------
+        ApplyMethod
+            The animation for the `Scene` object to `play` to change the mobject to
+            `DEFAULT_COLOR`.
+        """
+        return manim.FadeToColor(mobject, cls.DEFAULT_COLOR)
+
+    @classmethod
+    def highlight(cls, mobject: manim.VMobject) -> manim.ApplyMethod:
+        """
+        Parameters
+        ----------
+        mobject : VMobject
+            The mobject to set the `HIGHLIGHT_COLOR`.
+
+        Returns
+        -------
+        ApplyMethod
+            The animation for the `Scene` object to `play` to highlight the mobject.
+        """
+        return manim.FadeToColor(mobject, cls.HIGHLIGHT_COLOR)
+
+
+@dataclass(eq=False)
+class _ManimNode(manim.VGroup):
+    """
+    The `ManimNode` class represents a `Node` object in the `AGraph` object with manim.
+
+    The `ManimNode` class is a `VGroup`. For the convenience, it's also added with:
+
+    Parameters
+    ----------
+    shape : Dot | Circle | VGroup
+        - Dot for the nullnode (the start of the edge pointing the initial node)
+        - Circle for non-final state
+        - VGroup of a doublecircle for final state
+    label: Optional[Text]
+        The name of the state (N/A for the nullnode)
+    """
+
+    shape: manim.Dot | manim.Circle | manim.VGroup
+    label: Optional[manim.Text]
+
+    def __init__(self, node: pgv.Node) -> None:
+        """
+        Parameters
+        ----------
+        node : `pygraphviz.Node`
+            the node which to based on to construct a FA state.
+
+            `node`'s label is its `attr['label']` if given, otherwise its
+            `name` property. Its `attr` contains:
+            - 'fontsize': float str,
+            - 'height': float str,
+            - 'pos': f'{float str},{float str}',
+            - 'shape': 'point' or 'circle' or 'doublecircle',
+            - 'width': float str (possibly equals to 'height')
+        """
+        super().__init__(name=node.name)
+        radius = float(node.attr["height"]) / 2
+        if node.attr["shape"] == "point":
+            self.shape = Animate.default_init(manim.Dot)(radius=radius)
+            self.add(self.shape)
+        elif node.attr["shape"].endswith("circle"):
```

Comment:
Is there no third case? Should we raise an exception if the shape does not meet either of these two conditions?

Reply:
This is true, although the function can theoretically take in an argument of any type, even if by mistake. I'd prefer to have an exception raised if neither case is hit to avoid a weird noop if the shape isn't recognized.